server.go

```diff
@@ -1103,20 +1103,38 @@ func chainUnaryServerInterceptors(s *Server) {
 }
 
 func chainUnaryInterceptors(interceptors []UnaryServerInterceptor) UnaryServerInterceptor {
+	// the struct ensures the variables are allocated together, rather than separately, since we
+	// know they should be garbage collected together. This, together with the sync.Pool, saves
+	// 2 allocation and decreases time/call by about 25% on the microbenchmark.
+	type state struct {
+		i            int
+		next         UnaryHandler
+		info         *UnaryServerInfo
+		handler      UnaryHandler
+		interceptors []UnaryServerInterceptor
+	}
+	var statePool = sync.Pool{
```

Comment:
This seems like an unusual usage of `sync.Pool` (in that it's a local variable, though I guess it's still shared across the entire server).  Would this be better off as a global, though?

Reply:
I normally prefer avoiding globals when possible. But sure, this can be a global and it would still work (possibly even marginally better if you have multiple chains)